### PR TITLE
Editor state's refactoring - step 2/2

### PR DESCRIPTION
**Pull request by andresmoschini**
It is a big code refactoring, I would like to split it into several commits, but all changes are tied up, and untied them could take me more than a day.

This PR has three main aspects:

- Use React state in place of refs
- Split logic into several files
- Replace the _save_ promise

## Use React state in place of refs

When we were using refs to store the state, it was not easy to update the UI based on this because there was no React render associated and it was also difficult to test. 

It will be especially problematic to implement the UX improvements. For example, Unlayer's Programmatic Undo/Redo does not use React's state to show when there is any action in the queue to undo. We have to ask, and we cannot ask all the time, it depends on the state.

Also, testing the behavior is really complex and cannot be split into parts that _fit in a head_, we were forced to test it as a whole. Using React reducer state and effects, we can test each piece independently.

## Split logic into several files

We have split the functions and effects into many files with their own hooks each of them. I tried to keep functionality together based on the related behavior and the shared code, but actually, the division is arbitrary. I accept feedback.

## Replace the _save_ promise

Now, `smartSave` and `forceSave` are simply reducer's actions that throw a kind of saga with the asynchronous steps required to extract the content and save it. So, they cannot return a promise to know when the save is done.

On the other hand, the promise was not enough because what happen when there was more updates done after the _save request_?

Now, there is another reducer's action (`when-all-saved-action-requested`) that adds a function to the state to be executed when there are no pending updates to be saved.

It is a change in the behavior,  not only a refactoring.

### CODE REVIEW

**Comment by andresmoschini** [comment]
CC: @jcamposmk @ck-makingsense

File: src/components/singleton-editor/reducer.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  ) {
+    throw new Error(
+      "Unexpected scenario: savingDataCounter cannot be greater than updateCounter"
+    );
+  }
+  if (savedCounter > updateCounter) {
+    throw new Error(
+      "Unexpected scenario: savedDataCounter cannot be greater than updateCounter"
+    );
+  }
```
Comment (jcamposmk) [inline src/components/singleton-editor/reducer.ts:72]: No estoy muy seguro de esta parte. [Por aqui](https://stackoverflow.com/questions/62388219/is-it-safe-to-throw-an-error-in-a-react-reducer) hablan un poco al respecto indicando que debería chequearse antes. Otro dice que deberia modificar una propiedad indicando que hay un error. Nosé que piensas acerca de esto

Comment (andresmoschini) [reply src/components/singleton-editor/reducer.ts:72]: En realidad son escenarios inesperados, no deberían ocurrir salvo que haya un error en el código, entonces me gustaría hacerlos evidentes. 

Creo que modificar una propiedad indicando un error no sería correcto ya son escenarios teóricamente imposibles.

Igual voy a leer el artículo, y luego te digo.

Gracias!

Comment (andresmoschini) [reply src/components/singleton-editor/reducer.ts:72]: Leí el artículo y algunos de los links, estoy de acuerdo con ellos.

Pero entiendo que, en general, se refieren a excepciones que forman parte del flujo, por ejemplo errores de validación, que podrían ser capturadas y resueltas de alguna manera.

Las excepciones que yo puse, corresponden a escenarios imposibles que no puedo manejar con los tipos, no hay forma correcta de capturarlas, mi único interés es que si se da alguno de esos escenarios, se rompa cuanto antes de manera de identificar el problema.

Una alternativa sería usar simples `console.error`, pero entiendo que eso puede pasar desapercibido, en cambio la excepción va a generar claramente un error.

**Pull request merged**